File: src/Crawler/Analysis/Analyzer.php
<?php

/*
 * This file is part of the SiteOne Crawler.
 *
 * (c) Ján Regeš <jan.reges@siteone.cz>
 */

declare(strict_types=1);

namespace Crawler\Analysis;

use Crawler\Analysis\Result\UrlAnalysisResult;
use Crawler\Crawler;
use Crawler\ExtraColumn;
use Crawler\Options\Options;
use Crawler\Output\Output;
use Crawler\Result\Status;
use Crawler\Result\VisitedUrl;
use DOMDocument;

interface Analyzer
{

    /**
     * Set config for whole crawler with already parsed options
     * @param Options $options
     * @return void
     */
    public function setConfig(Options $options): void;

    /**
     * Set output for this exporter to show progress
     * @param Output $output
     * @return void
     */
    public function setOutput(Output $output): void;

    /**
     * Set crawler to get data from it
     * @param Crawler $crawler
     * @return void
     */
    public function setCrawler(Crawler $crawler): void;

    /**
     * Set status to get data from it
     * @param Status $status
     * @return void
     */
    public function setStatus(Status $status): void;

    /**
     * Do your analysis and set results to output
     * @return void
     */
    public function analyze(): void;

    /**
     * Do your analysis right now visited URL. Body and headers are already downloaded and decompressed.
     * Return null if you don't want to analyze this URL, otherwise return UrlAnalysisResult with your results.
     *
     * @param VisitedUrl $visitedUrl
     * @param string|null $body
     * @param DOMDocument|null $dom
     * @param array|null $headers
     * @return UrlAnalysisResult|null
     */
    public function analyzeVisitedUrl(VisitedUrl $visitedUrl, ?string $body, ?DOMDocument $dom, ?array $headers): ?UrlAnalysisResult;

    /**
     * If you want to show URL analysis results in table column (as numbers with severity icons), return name of column
     * under which your results will be shown. This column will be added to the table automatically.
     *
     * @return ExtraColumn|null
     */
    public function showAnalyzedVisitedUrlResultAsColumn(): ?ExtraColumn;

    /**
     * Should this analyzer be activated based on options?
     * @return bool
     */
    public function shouldBeActivated(): bool;

    /**
     * Get order of this analyzer (lower = earlier)
     * @return int
     */
    public function getOrder(): int;

    /**
     * Get options for this analyzer
     * @return Options
     */
    public static function getOptions(): Options;

}

File: src/analysis/analyzer.rs
// SiteOne Crawler - Analyzer trait
// (c) Jan Reges <jan.reges@siteone.cz>

use std::collections::HashMap;

use crate::analysis::result::url_analysis_result::UrlAnalysisResult;
use crate::extra_column::ExtraColumn;
use crate::result::visited_url::VisitedUrl;

/// Trait that all analyzers must implement.
pub trait Analyzer: Send + Sync {
    /// Do your analysis and set results to output (post-crawl).
    /// Called after all URLs have been visited.
    fn analyze(&mut self, status: &crate::result::status::Status, output: &mut dyn crate::output::output::Output);

    /// Do your analysis for a just-visited URL.
    /// Body and headers are already downloaded and decompressed.
    /// Return None if you don't want to analyze this URL,
    /// otherwise return UrlAnalysisResult with your results.
    fn analyze_visited_url(
        &mut self,
        _visited_url: &VisitedUrl,
        _body: Option<&str>,
        _headers: Option<&HashMap<String, String>>,
    ) -> Option<UrlAnalysisResult> {
        None
    }

    /// If you want to show URL analysis results in table column,
    /// return the ExtraColumn under which results will be shown.
    fn show_analyzed_visited_url_result_as_column(&self) -> Option<ExtraColumn> {
        None
    }

    /// Should this analyzer be activated based on options?
    fn should_be_activated(&self) -> bool;

    /// Get order of this analyzer (lower = earlier).
    fn get_order(&self) -> i32;

    /// Get the name of this analyzer.
    fn get_name(&self) -> &str;

    /// Get execution times of analyzer methods.
    fn get_exec_times(&self) -> &HashMap<String, f64>;

    /// Get execution counts of analyzer methods.
    fn get_exec_counts(&self) -> &HashMap<String, usize>;
}
